pvanalytics/features/daytime.py

```diff
@@ -0,0 +1,210 @@
+    - near-zero rolling median at the same time over the surrounding
+      week
+
+    It is possible that mid-day times where power goes near zero or
+    stops changing can be incorrectly classified as night. To correct
+    these errors times where the total duration of periods marked
+    night or day is too long or too short are identified and values at
+    these times are compared to values at the same time for the
+    preceding two weeks and following two weeks. The day/night value
+    is adjusted to match the median value of these days.
+
+    Finally any values that are True in `clipping` are marked as day.
+
+    Parameters
+    ----------
+    series : Series
+        Time series of power or irradiance.
+    outliers : Series, optional
+        Boolean time series with True for values in `series` that are
+        outliers.
+    low_value_threshold : float, default 0.003
+        Maximum normalized value for a time to be considered night.
+    low_median_threshold : float, default 0.0015
+        Minimum rolling median for a time to be considered night.
+    low_diff_threshold : float, default 0.0005
+        Maximum derivative for a time to be considered night.
+    clipping : Series, optional
+        True when clipping indicated. Any values where clipping is
+        indicated are automatically considered 'daytime'.
+
+    Returns
+    -------
+    Series
+        Boolean time series with True for times that are during the
+        day.
+
+    Notes
+    -----
+
+    ``NA`` values are treated like zeros.
+
+    Derived from the PVFleets QA Analysis project.
+
+    """
+    series = series.fillna(value=0)
+    series_norm = _filter_and_normalize(series, outliers).fillna(value=0)
+    minutes_per_value = _freqstr_to_minutes(
+        freq or pd.infer_freq(series.index)
+    )
+    first_order_diff = series_norm.diff() / minutes_per_value
+    rolling_median = _rolling_by_minute(
+        series_norm,
+        days=7,
+        f=pd.core.window.RollingGroupby.median
+    )
+
+    # Night-time if two of the following are satisfied:
+    # - Near-zero value
+    # - Near-zero first-order derivative
+    # - Near-zero rolling median
+    low_value = series_norm <= low_value_threshold
+    low_median = rolling_median <= low_median_threshold
+    low_diff = abs(first_order_diff) <= low_diff_threshold
+
+    night = ((low_value & low_diff)
+             | (low_value & low_median)
+             | (low_diff & low_median))
+    # Fix erroneous classifications (e.g. midday outages where power
+    # goes to 0 and stays there for several hours, clipping classified
+    # as night, and night-time periods that are too long)
+    night_corrected_midday = _filter_midday_errors(night, minutes_per_value)
+    night_corrected_clipping = ~((clipping or False)
+                                 | (~night_corrected_midday))
+    night_corrected_edges = _filter_edge_of_day_errors(
+        night_corrected_clipping, minutes_per_value
+    )
```

Comment:
A few thoughts:

- Rename these functions "filter" isn't the right word. Let's call them `_correct_*_errors()`.
- Add a comment describing why we call `_filter_edge_of_day_errors()` and maybe also why that function is separate from `_filter_midday_errors()`
- Should the clipping correction come first? It seems like that would give more information to the two error correction algorithms.

Reply:
Is there any value in making any of the correction steps optional? (e.g. pass Booleans `correct_midday`, and `correct_edge_of_day`)